src/crimson/osd/backfill_state.cc

```diff
@@ -0,0 +1,523 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "crimson/osd/backfill_state.h"
+#include "crimson/osd/backfill_facades.h"
+#include "crimson/osd/pg.h"
+#include "osd/PeeringState.h"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_osd);
+  }
+}
+
+namespace crimson::osd {
+
+BackfillState::BackfillState(
+  BackfillState::BackfillListener& backfill_listener,
+  std::unique_ptr<BackfillState::PeeringFacade> peering_state,
+  std::unique_ptr<BackfillState::PGFacade> pg)
+  : backfill_machine(*this,
+                     backfill_listener,
+                     std::move(peering_state),
+                     std::move(pg)),
+    progress_tracker(
+      std::make_unique<BackfillState::ProgressTracker>(backfill_machine))
+{
+  logger().debug("{}:{}", __func__, __LINE__);
+  backfill_machine.initiate();
+}
+
+BackfillState::~BackfillState() = default;
+
+BackfillState::BackfillMachine::BackfillMachine(
+  BackfillState& backfill_state,
+  BackfillState::BackfillListener& backfill_listener,
+  std::unique_ptr<BackfillState::PeeringFacade> peering_state,
+  std::unique_ptr<BackfillState::PGFacade> pg)
+  : backfill_state(backfill_state),
+    backfill_listener(backfill_listener),
+    peering_state(std::move(peering_state)),
+    pg(std::move(pg))
+{}
+
+BackfillState::BackfillMachine::~BackfillMachine() = default;
+
+BackfillState::Initial::Initial(my_context ctx)
+  : my_base(ctx)
+{
+  bs().last_backfill_started = ps().earliest_backfill();
+  logger().debug("{}: bft={} from {}",
+                 __func__, ps().get_backfill_targets(),
+                 bs().last_backfill_started);
+  for (const auto& bt : ps().get_backfill_targets()) {
+    logger().debug("{}: target shard {} from {}",
+                   __func__, bt, ps().get_peer_info(bt).last_backfill);
+  }
+  ceph_assert(ps().get_backfill_targets().size());
+  ceph_assert(!bs().last_backfill_started.is_max());
+}
+
+boost::statechart::result
+BackfillState::Initial::react(const BackfillState::Trigerred& evt)
+{
+  logger().debug("{}: backfill triggered", __func__);
+  ceph_assert(bs().last_backfill_started == ps().earliest_backfill());
+  // initialize BackfillIntervals
+  for (const auto& bt : ps().get_backfill_targets()) {
+    bs().peer_backfill_info[bt].reset(
+      ps().get_peer_info(bt).last_backfill);
+  }
+  bs().backfill_info.reset(bs().last_backfill_started);
+  bs().backfills_in_flight.clear();
+  bs().pending_backfill_updates.clear();
+  if (Enqueuing::all_enqueued(ps(),
+                              bs().backfill_info,
+                              bs().peer_backfill_info)) {
```

Comment:
Just for clarity, if all_enqueued() here, then we've actually completed our pushes, right (otherwise we'd be in Waiting).  Can we actually get here under those circumstances?

Reply:
This is executed only once – after the initialization of entire FSM. The `peer_backfill_info` map, which `all_enqueued()` depends on, is updated with `pg_info_t::last_backfill` and that's why I'm little bit paranoiac here.